### PR TITLE
Add PDB e2e test

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/openshift/openshift-azure/issues/309

### CODE REVIEW

File: test/e2e/osa_test.go [hunk 1]
```diff
@@ -83,4 +129,23 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test",
+			},
+			Spec: policy.PodDisruptionBudgetSpec{
+				MaxUnavailable: &maxUnavailable,
+				Selector:       selector,
+			},
+		}
+
+		_, err = kubeClient(c.cs.Config.AdminKubeconfig).Policy().PodDisruptionBudgets("e2e-tests").Create(pdb)
```
Comment (mjudeikis) [inline test/e2e/osa_test.go]: Is this `system:admin` for now? Not yet delegated?

Comment (kargakis) [reply test/e2e/osa_test.go]: Yes, there is a TODO in kubeClient to fix that

File: test/e2e/osa_test.go [hunk 2]
```diff
@@ -30,10 +34,52 @@ var _ = BeforeSuite(func() {
...
+	// to test with different permissions.
+	c.kc, err = managedcluster.ClientsetFromConfig(c.cs.Config.AdminKubeconfig)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Create a namespace for running tests and assign osadmin
+	// as a project admin to it. Delegate permissions to osadmin
+	// to allow us to run tests in a non-privileged context.
+	_, err = c.kc.CoreV1().Namespaces().Create(&v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "e2e-tests",
```
Comment (mjudeikis) [inline test/e2e/osa_test.go]: are we using one namespace for all tests? if we introduce parallelism to the tests will it not clash?

Comment (mjudeikis) [reply test/e2e/osa_test.go]: I have seen this now in origin conformance testing. If in example few tests use same namespace and one of them is testing "network policy" or rbac restriction it automatically can flake another test in the same namespace, despite the fact it uses different resources.

Comment (kargakis) [reply test/e2e/osa_test.go]: > can flake another test in the same namespace, despite the fact it uses different resources.

Can you elaborate? Is it namespace cleanup that triggers the flake or something else?

Comment (mjudeikis) [reply test/e2e/osa_test.go]: that could be too. I think each test case should create unique namespace for itself and clean after.

Comment (kargakis) [reply test/e2e/osa_test.go]: I am against namespace cleanup. It's probably one of the flakiest things I have ever seen in origin e2e.

**Comment by kargakis** [comment]
@mjudeikis @jim-minter @pweil- ptal, this is now ready to get merged

**Comment by jim-minter** [comment]
I feel like a few things are problematic here:
* I think we should start down the line of being able to run selected tests straight with an end user kubeconfig - if you solve the associated problems, others can build tests in your footsteps.  Perhaps we should have two separate test suites - one expecting system:admin, one not?
* I want to avoid building stuff assuming the existence of `osadmin` or what its associated [cluster]rolebindings might be
* I agree with MJ about separate namespaces and at least trying to clean them up

**Comment by mjudeikis** [comment]
> Perhaps we should have two separate test suites - one expecting system:admin, one not?

I think we need 3:
* Project admin
* System:Admin
* (future) azure-cluster-admin. And considering gingko uses tags to whitelist test, should we add relevant tagging to tests too? based on scope, type, etc

**Comment by kargakis** [comment]
>  I think we should start down the line of being able to run selected tests straight with an end user kubeconfig - if you solve the associated problems, others can build tests in your footsteps. Perhaps we should have two separate test suites - one expecting system:admin, one not?

Do we need to precreate the end-user kubeconfig as part of cluster creation then? How do we bind it to users as they are getting access to the cluster?

> I want to avoid building stuff assuming the existence of osadmin or what its associated [cluster]rolebindings might be

I don't think the name of the user matters and should be easily swapped for whatever we want to come up with. The important part here is the binding to project admin.

> I agree with MJ about separate namespaces and at least trying to clean them up

Namespace cleanup, as I mentioned already, used to be the biggest source of flakes in Origin conformance tests. If we want to test namespace cleanup, that should be a scoped test. There is no other reason to do namespace cleanup if we are going to delete the cluster once the e2e tests are done. Separate namespace per test is fine.

**Comment by kargakis** [comment]
> There is no other reason to do namespace cleanup if we are going to delete the cluster once the e2e tests are done. Separate namespace per test is fine.

I guess we want to do namespace cleanup if we plan to run these tests in production clusters :(

**Comment by jim-minter** [comment]
> I don't think the name of the user matters and should be easily swapped for whatever we want to come up with. The important part here is the binding to project admin.

I think I got a bit confused because grafana in azure-misc was making osadmin a cluster reader.  It doesn't do that any more.  As long as we don't pull those kinds of tricks, using osadmin works for now.

**Comment by kargakis** [comment]
Opened https://github.com/openshift/openshift-azure/pull/528 for splitting out the e2e framework refactoring.

**Comment by kargakis** [comment]
@mjudeikis @jim-minter this should be ready now (I hope)

**Comment by kargakis** [comment]
The tests wont pass until https://github.com/openshift/openshift-azure/pull/545 gets merged.

File: test/e2e/client_test.go [hunk 3]
```diff
@@ -40,21 +44,38 @@ func newTestClient(kubeconfig string) *testClient {
...
+		}}); err != nil {
+		return err
+	}
 	t.namespace = namespace
 	// TODO: Wait for a successful SAR check
+	return nil
 }
 
-func (t *testClient) cleanupNamespace(timeout time.Duration) {
+func (t *testClient) cleanupNamespace(timeout time.Duration) error {
```
Comment (jim-minter) [inline test/e2e/client_test.go]: nit: timeout not used

File: test/e2e/client_test.go [hunk 4]
```diff
@@ -40,21 +44,38 @@ func newTestClient(kubeconfig string) *testClient {
...
+func (t *testClient) createNamespace(namespace string) error {
+	if _, err := t.pc.ProjectRequests().Create(&project.ProjectRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		}}); err != nil {
+		return err
+	}
 	t.namespace = namespace
 	// TODO: Wait for a successful SAR check
+	return nil
```
Comment (jim-minter) [inline test/e2e/client_test.go]: may just be worth putting in some code that waits for the obvious service accounts to be created, before that bites a new starter

File: test/e2e/client_test.go [hunk 5]
```diff
@@ -40,21 +44,38 @@ func newTestClient(kubeconfig string) *testClient {
...
 }
 
-func (t *testClient) cleanupNamespace(timeout time.Duration) {
+func (t *testClient) cleanupNamespace(timeout time.Duration) error {
 	if t.namespace == "" {
-		return
+		return nil
```
Comment (jim-minter) [inline test/e2e/client_test.go]: nit: simplifying this and have people to defer t.cleanupNamespace() is tempting, but it's not a blocker

Comment (kargakis) [reply test/e2e/client_test.go]: This is actually part of a AfterEach function so any new tests added should not care about namespace cleanup as it's handled automatically for them.

**Pull request merged**